presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java

```diff
@@ -140,16 +145,43 @@ private Patterns() {}
         return typeOf(WindowNode.class);
     }
 
-    public static <C> Property<PlanNode, C, PlanNode> source()
-    {
-        return optionalProperty("source", node -> node.getSources().size() == 1 ?
-                Optional.of(node.getSources().get(0)) :
-                empty());
-    }
-
-    public static <C> Property<PlanNode, C, List<PlanNode>> sources()
-    {
-        return property("sources", PlanNode::getSources);
+    public static Explore<PlanNode, Rule.Context, PlanNode> source()
+    {
+        return explore(
+                "source",
+                (node, context) -> {
+                    if (node.getSources().size() != 1) {
+                        return Stream.of();
+                    }
+                    return context.getLookup().resolveGroup(getOnlyElement(node.getSources()));
+                });
+    }
+
+    public static Explore<PlanNode, Rule.Context, List<PlanNode>> sources()
+    {
+        return explore(
+                "sources",
+                (PlanNode node, Rule.Context context) -> {
+                    List<Stream<PlanNode>> sourceStreams = node.getSources().stream()
+                            .map(source -> context.getLookup().resolveGroup(source))
+                            .collect(toImmutableList());
+
+                    return Streams.stream(new AbstractIterator<List<PlanNode>>() {
+                        @Override
+                        protected List<PlanNode> computeNext()
+                        {
+                            List<PlanNode> sources = sourceStreams.stream()
+                                    .map(source -> source.findFirst()
```

Comment:
Once you call `findFirst`, the stream should not be used anymore.
Maybe convert to `List<Iterator>`?